File: bindings/lua/examples/manual_types.lua
local source = debug.getinfo(1, "S").source:sub(2)
local dir = source:match("^(.*)/[^/]+$") or "."
package.path = dir .. "/../?.lua;" .. package.path

local urb = require("urb_ffi")
local ffi, mem = urb.ffi, urb.memory

local libc = ffi.open("libc.so.6")
local puts_type = ffi.type.func(ffi.type.i32(), { ffi.type.cstring() }, { name = "puts" })
local puts = ffi.bind(ffi.sym(libc, "puts"), puts_type)

puts("hello from urb-ffi manual type descriptors (lua)")

local int_ptr = mem.alloc(4)
local int_ref = ffi.global(int_ptr, ffi.type.i32())
int_ref.value = 123
print("global i32 =", int_ref.value)

local point_type = ffi.type.struct({
    { name = "x", type = ffi.type.i32() },
    { name = "y", type = ffi.type.i32() },
})
local point_ptr = mem.alloc(ffi.type.sizeof(point_type))
local point = ffi.global(point_ptr, point_type)
point.x = 7
point.y = 11
print("point =", point.x, point.y)

mem.free(point_ptr)
mem.free(int_ptr)
ffi.close(libc)

File: bindings/node/examples/manual_types.js
'use strict';

const { ffi, memory: mem } = require('..');

const libcName = process.platform === 'darwin'
	? '/usr/lib/libSystem.B.dylib'
	: process.platform === 'win32'
		? 'ucrtbase.dll'
		: 'libc.so.6';

const libc = ffi.open(libcName, ffi.flags.NOW | ffi.flags.LOCAL);
const putsType = ffi.type.func(ffi.type.i32(), [ffi.type.cstring()], { name: 'puts' });
const puts = ffi.bind(ffi.sym(libc, 'puts'), putsType);

puts('hello from urb-ffi manual type descriptors (node)');

const intPtr = mem.alloc(4);
const intRef = ffi.global(intPtr, ffi.type.i32());
intRef.value = 123;
console.log('global i32 =', intRef.value);

const pointType = ffi.type.struct([
	{ name: 'x', type: ffi.type.i32() },
	{ name: 'y', type: ffi.type.i32() },
]);
const pointPtr = mem.alloc(ffi.type.sizeof(pointType));
const point = ffi.global(pointPtr, pointType);
point.x = 7;
point.y = 11;
console.log('point =', point.x, point.y);

mem.free(pointPtr);
mem.free(intPtr);
ffi.close(libc);